### ISSUE TITLE
Question: Source -> Transform -> Merge

### ISSUE THREAD

**Issue by sliceofbytes** [roadrunnerlenny/etlbox#54]
Source -> Transform -> Merge

Is this a flow that is supported. Even in simple cases I'm getting exception:

`ArgumentNullException: Value cannot be null. (Parameter 'source')`

My simple test case looks like this.

```csharp
	public class Name
	{
               public string ID { get; set; }

               public string FIRST_NAME { get; set; }

               public string LAST_NAME { get; set; }
        }
```

```csharp
	public class People : MergeableRow
	{
		[CompareColumn]
		public string FirstName { get; set; } //20

		[CompareColumn]
		public string LastName { get; set; } //30

		[IdColumn]
		public string Id { get; set; } //10
	}
```

```csharp
	class Program
	{

		private static SqlConnectionManager _sourceConnection;
		private static SqlConnectionManager _destinationConnection;

		static void Main(string[] args)
		{

			var builder = new ConfigurationBuilder()
	                        .SetBasePath(Directory.GetCurrentDirectory())
	                        .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true);

			IConfigurationRoot configuration = builder.Build();

			_sourceConnection = new SqlConnectionManager(configuration.GetConnectionString("Source"));
			_destinationConnection = new SqlConnectionManager(configuration.GetConnectionString("Destination"));

			var nameSource = new DbSource<Name>(_sourceConnection, "Name");
			var personMerge = new DbMerge<People>(_destinationConnection, "People");
			

			var transform = new RowTransformation<Name, People>(d =>
			{
				return new People()
				{
					FirstName = d.FIRST_NAME,
					LastName = d.LAST_NAME,
					Id = d.ID
				};
			});

			nameSource.LinkTo(transform);
			transform.LinkTo(personMerge);

			nameSource.Execute();
			personMerge.Wait();

			Console.WriteLine("Complete");

		}
```

**Comment by roadrunnerlenny** [comment]
Hi @sliceofbytes ,
thank you for your feedback.
Actually, the dataflow you created with ETLBox should work just fine. 
I just added a test which basically simulates the exact same flow:
https://github.com/roadrunnerlenny/etlbox/blob/b8326f4dab74067801de210b54e953801a6058f5/TestsETLBox/src/DataFlowTests/DBMerge/DbMergeAcrossDatabasesTests.cs#L75-L92
The tests worked perfectly in my test environment.

Which version of ETLBox are you using? 

I looked in your code, I actually would suspect that the error is thrown in this line
`_sourceConnection = new SqlConnectionManager(configuration.GetConnectionString("Source"));`

The `configuration.GetConnectionString("Source")` probably tries to access some configuration file, and perhaps this is not accessable or does not contain the configuration you are looking for? Could you try something like this:

```C#
string _connectionString = configuration.GetConnectionString("Source");
_sourceConnection = new SqlConnectionManager(_connectionString);
```
Can your program successfully retrieve the connection string and store it in `_connectionString`?

**Comment by sliceofbytes** [comment]
Yes,I confirmed ETLBox was working with a simple Source -> Transform -> Destination.

I only ran into problems when trying to test out the merge feature.

Changing the code I posted above to a DbDestination instead of DbMerge works correctly with my data set.

**Comment by roadrunnerlenny** [comment]
Hi,
I reproduced your code in my test environment and it worked fine using the DbMerge. 
So probably the difference lays in the underlying data? Could you give me an example data set to reproduce the error?
What version are you using?

**Comment by sliceofbytes** [comment]
I'm using 1.8.0-rc2, I've been able to recreate the issue with sanitized data so give me a few moments to put a repro together and I can share everything.

**Comment by sliceofbytes** [comment]
Here is a repo with the code and the scripts to create both tables and populate the source data. 

https://github.com/sliceofbytes/ETLBoxRepro

**Comment by roadrunnerlenny** [comment]
Hi,
thanks for example repo! Now I get the exact same error in my environment. I will have a deeper look and hopefully can come up with a bugfix pretty soon. I will keep you updated!

**Comment by sliceofbytes** [comment]
Thanks!

I've added a sqlite example and got the same error so it seems unrelated to the db instance.

https://github.com/sliceofbytes/ETLBoxRepro/tree/sqlite

**Comment by roadrunnerlenny** [comment]
Hi @sliceofbytes ,

I just released version 1.8.0 (https://www.nuget.org/packages/ETLBox/1.8.0) that contains a fix for this issue. Let me know if this works for you.

**Comment by sliceofbytes** [comment]
I can confirm this fixed the original bug. I can now DBMerge the dataset in question. It is extremely slow though. Taking 56 minutes to DbMerge 156k rows.

**Comment by roadrunnerlenny** [comment]
Thanks for the feedback. That really is way to slow. I need to setup some test to reproduce this, and I will try to create a fix for this. 

Just a quick check to see if this is related to the deletion or updating of existing records: did you execute the merge on an empty destination or was there already a lot of data in the destination that potentially could get updated? Could you try the same run with „UseTruncateMethod“ = true?

**Comment by sliceofbytes** [comment]
I had an equal number of rows in the destination 156k. So there are lots of rows that could potentially get updated (Only 1 actually had a change). I'm trying with `UseTruncateMethod = true`. It appears to be equally slow but I'll let it run out and give you the exact time.

**Comment by sliceofbytes** [comment]
It seems to progressively get slower as it processes the data.

```
2020-02-25 10:01:01.8145|INFO|ETL|Write data into table PeopleTest processed 1000 records.
2020-02-25 10:01:02.1826|INFO|ETL|Lookup processed 2000 records.
2020-02-25 10:01:02.1921|INFO|ETL|Write data into table PeopleTest processed 2000 records.
2020-02-25 10:01:02.9330|INFO|ETL|Lookup processed 3000 records.
2020-02-25 10:01:02.9437|INFO|ETL|Write data into table PeopleTest processed 3000 records.
2020-02-25 10:01:03.9192|INFO|ETL|Lookup processed 4000 records.
2020-02-25 10:01:03.9294|INFO|ETL|Write data into table PeopleTest processed 4000 records.
2020-02-25 10:01:05.2385|INFO|ETL|Lookup processed 5000 records.
2020-02-25 10:01:05.2385|INFO|ETL|Write data into table PeopleTest processed 5000 records.
2020-02-25 10:01:06.8226|INFO|ETL|Lookup processed 6000 records.
2020-02-25 10:01:06.8318|INFO|ETL|Write data into table PeopleTest processed 6000 records.
2020-02-25 10:01:08.7133|INFO|ETL|Lookup processed 7000 records.
2020-02-25 10:01:08.7285|INFO|ETL|Write data into table PeopleTest processed 7000 records.
2020-02-25 10:01:10.8858|INFO|ETL|Lookup processed 8000 records.
2020-02-25 10:01:10.8963|INFO|ETL|Write data into table PeopleTest processed 8000 records.
2020-02-25 10:01:13.3056|INFO|ETL|Lookup processed 9000 records.
2020-02-25 10:01:13.3168|INFO|ETL|Write data into table PeopleTest processed 9000 records.
2020-02-25 10:01:15.9817|INFO|ETL|Lookup processed 10000 records.
2020-02-25 10:01:15.9918|INFO|ETL|Write data into table PeopleTest processed 10000 records.
2020-02-25 10:01:19.0339|INFO|ETL|Lookup processed 11000 records.
2020-02-25 10:01:19.0460|INFO|ETL|Write data into table PeopleTest processed 11000 records.
2020-02-25 10:01:22.4375|INFO|ETL|Lookup processed 12000 records.
2020-02-25 10:01:22.4375|INFO|ETL|Write data into table PeopleTest processed 12000 records.
2020-02-25 10:01:26.4786|INFO|ETL|Lookup processed 13000 records.
2020-02-25 10:01:26.4906|INFO|ETL|Write data into table PeopleTest processed 13000 records.
2020-02-25 10:01:30.5136|INFO|ETL|Lookup processed 14000 records.
2020-02-25 10:01:30.5264|INFO|ETL|Write data into table PeopleTest processed 14000 records.
2020-02-25 10:01:34.7488|INFO|ETL|Lookup processed 15000 records.
2020-02-25 10:01:34.7619|INFO|ETL|Write data into table PeopleTest processed 15000 records.
2020-02-25 10:01:39.2682|INFO|ETL|Lookup processed 16000 records.
2020-02-25 10:01:39.2682|INFO|ETL|Write data into table PeopleTest processed 16000 records.
2020-02-25 10:01:44.0474|INFO|ETL|Lookup processed 17000 records.
2020-02-25 10:01:44.0974|INFO|ETL|Write data into table PeopleTest processed 17000 records.
2020-02-25 10:01:49.1670|INFO|ETL|Lookup processed 18000 records.
2020-02-25 10:01:49.1803|INFO|ETL|Write data into table PeopleTest processed 18000 records.
2020-02-25 10:01:54.5174|INFO|ETL|Lookup processed 19000 records.
2020-02-25 10:01:54.5327|INFO|ETL|Write data into table PeopleTest processed 19000 records.
2020-02-25 10:02:00.3479|INFO|ETL|Lookup processed 20000 records.
2020-02-25 10:02:00.3839|INFO|ETL|Write data into table PeopleTest processed 20000 records.
2020-02-25 10:02:06.7383|INFO|ETL|Lookup processed 21000 records.
2020-02-25 10:02:06.7527|INFO|ETL|Write data into table PeopleTest processed 21000 records.
2020-02-25 10:02:13.7300|INFO|ETL|Lookup processed 22000 records.
2020-02-25 10:02:13.7431|INFO|ETL|Write data into table PeopleTest processed 22000 records.
2020-02-25 10:02:20.3402|INFO|ETL|Lookup processed 23000 records.
2020-02-25 10:02:20.3540|INFO|ETL|Write data into table PeopleTest processed 23000 records.
2020-02-25 10:02:27.4014|INFO|ETL|Lookup processed 24000 records.
2020-02-25 10:02:27.4160|INFO|ETL|Write data into table PeopleTest processed 24000 records.
2020-02-25 10:02:35.4459|INFO|ETL|Lookup processed 25000 records.
2020-02-25 10:02:35.4593|INFO|ETL|Write data into table PeopleTest processed 25000 records.
2020-02-25 10:02:43.2697|INFO|ETL|Lookup processed 26000 records.
2020-02-25 10:02:43.2860|INFO|ETL|Write data into table PeopleTest processed 26000 records.
2020-02-25 10:02:51.0563|INFO|ETL|Lookup processed 27000 records.
2020-02-25 10:02:51.0721|INFO|ETL|Write data into table PeopleTest processed 27000 records.
2020-02-25 10:02:58.9626|INFO|ETL|Lookup processed 28000 records.
2020-02-25 10:02:58.9779|INFO|ETL|Write data into table PeopleTest processed 28000 records.
2020-02-25 10:03:07.2788|INFO|ETL|Lookup processed 29000 records.
2020-02-25 10:03:07.2943|INFO|ETL|Write data into table PeopleTest processed 29000 records.
2020-02-25 10:03:15.4967|INFO|ETL|Lookup processed 30000 records.
2020-02-25 10:03:15.5152|INFO|ETL|Write data into table PeopleTest processed 30000 records.
2020-02-25 10:03:24.0681|INFO|ETL|Lookup processed 31000 records.
2020-02-25 10:03:24.0949|INFO|ETL|Write data into table PeopleTest processed 31000 records.
2020-02-25 10:03:33.1424|INFO|ETL|Lookup processed 32000 records.
2020-02-25 10:03:33.1595|INFO|ETL|Write data into table PeopleTest processed 32000 records.
2020-02-25 10:03:42.9968|INFO|ETL|Lookup processed 33000 records.
2020-02-25 10:03:43.0138|INFO|ETL|Write data into table PeopleTest processed 33000 records.
2020-02-25 10:03:52.8166|INFO|ETL|Lookup processed 34000 records.
2020-02-25 10:03:52.8353|INFO|ETL|Write data into table PeopleTest processed 34000 records.
2020-02-25 10:04:02.7730|INFO|ETL|Lookup processed 35000 records.
2020-02-25 10:04:02.7893|INFO|ETL|Write data into table PeopleTest processed 35000 records.
2020-02-25 10:04:13.0552|INFO|ETL|Lookup processed 36000 records.
2020-02-25 10:04:13.0715|INFO|ETL|Write data into table PeopleTest processed 36000 records.
2020-02-25 10:04:23.4962|INFO|ETL|Lookup processed 37000 records.
2020-02-25 10:04:23.5127|INFO|ETL|Write data into table PeopleTest processed 37000 records.
2020-02-25 10:04:34.3497|INFO|ETL|Lookup processed 38000 records.
2020-02-25 10:04:34.3666|INFO|ETL|Write data into table PeopleTest processed 38000 records.
2020-02-25 10:04:45.9643|INFO|ETL|Lookup processed 39000 records.
2020-02-25 10:04:45.9890|INFO|ETL|Write data into table PeopleTest processed 39000 records.
2020-02-25 10:04:58.3750|INFO|ETL|Lookup processed 40000 records.
2020-02-25 10:04:58.3924|INFO|ETL|Write data into table PeopleTest processed 40000 records.
2020-02-25 10:05:10.3049|INFO|ETL|Lookup processed 41000 records.
2020-02-25 10:05:10.3298|INFO|ETL|Write data into table PeopleTest processed 41000 records.
2020-02-25 10:05:22.5117|INFO|ETL|Lookup processed 42000 records.
2020-02-25 10:05:22.5343|INFO|ETL|Write data into table PeopleTest processed 42000 records.
2020-02-25 10:05:35.1411|INFO|ETL|Lookup processed 43000 records.
2020-02-25 10:05:35.1621|INFO|ETL|Write data into table PeopleTest processed 43000 records.
2020-02-25 10:05:49.4876|INFO|ETL|Lookup processed 44000 records.
2020-02-25 10:05:49.5079|INFO|ETL|Write data into table PeopleTest processed 44000 records.
2020-02-25 10:06:03.6134|INFO|ETL|Lookup processed 45000 records.
2020-02-25 10:06:03.6339|INFO|ETL|Write data into table PeopleTest processed 45000 records.
2020-02-25 10:06:17.7252|INFO|ETL|Lookup processed 46000 records.
2020-02-25 10:06:17.7456|INFO|ETL|Write data into table PeopleTest processed 46000 records.
2020-02-25 10:06:32.6070|INFO|ETL|Lookup processed 47000 records.
2020-02-25 10:06:32.6453|INFO|ETL|Write data into table PeopleTest processed 47000 records.
2020-02-25 10:06:47.3548|INFO|ETL|Lookup processed 48000 records.
2020-02-25 10:06:47.3768|INFO|ETL|Write data into table PeopleTest processed 48000 records.
2020-02-25 10:07:03.8191|INFO|ETL|Lookup processed 49000 records.
2020-02-25 10:07:03.8407|INFO|ETL|Write data into table PeopleTest processed 49000 records.
2020-02-25 10:07:18.8187|INFO|ETL|Lookup processed 50000 records.
2020-02-25 10:07:18.8412|INFO|ETL|Write data into table PeopleTest processed 50000 records.
2020-02-25 10:07:36.6366|INFO|ETL|Lookup processed 51000 records.
2020-02-25 10:07:36.6560|INFO|ETL|Write data into table PeopleTest processed 51000 records.
2020-02-25 10:07:53.4701|INFO|ETL|Lookup processed 52000 records.
2020-02-25 10:07:53.5032|INFO|ETL|Write data into table PeopleTest processed 52000 records.
2020-02-25 10:08:10.5839|INFO|ETL|Lookup processed 53000 records.
2020-02-25 10:08:10.6165|INFO|ETL|Write data into table PeopleTest processed 53000 records.
2020-02-25 10:08:28.0946|INFO|ETL|Lookup processed 54000 records.
2020-02-25 10:08:28.1162|INFO|ETL|Write data into table PeopleTest processed 54000 records.
2020-02-25 10:08:45.2279|INFO|ETL|Lookup processed 55000 records.
2020-02-25 10:08:45.2506|INFO|ETL|Write data into table PeopleTest processed 55000 records.
2020-02-25 10:09:02.1729|INFO|ETL|Lookup processed 56000 records.
2020-02-25 10:09:02.1952|INFO|ETL|Write data into table PeopleTest processed 56000 records.
2020-02-25 10:09:19.4805|INFO|ETL|Lookup processed 57000 records.
2020-02-25 10:09:19.5052|INFO|ETL|Write data into table PeopleTest processed 57000 records.
2020-02-25 10:09:36.5400|INFO|ETL|Lookup processed 58000 records.
2020-02-25 10:09:36.5685|INFO|ETL|Write data into table PeopleTest processed 58000 records.
2020-02-25 10:09:54.3756|INFO|ETL|Lookup processed 59000 records.
2020-02-25 10:09:54.4016|INFO|ETL|Write data into table PeopleTest processed 59000 records.
2020-02-25 10:10:11.6670|INFO|ETL|Lookup processed 60000 records.
2020-02-25 10:10:11.6938|INFO|ETL|Write data into table PeopleTest processed 60000 records.
2020-02-25 10:10:30.5379|INFO|ETL|Lookup processed 61000 records.
2020-02-25 10:10:30.5608|INFO|ETL|Write data into table PeopleTest processed 61000 records.
2020-02-25 10:10:48.3510|INFO|ETL|Lookup processed 62000 records.
2020-02-25 10:10:48.3741|INFO|ETL|Write data into table PeopleTest processed 62000 records.
2020-02-25 10:11:08.3678|INFO|ETL|Lookup processed 63000 records.
2020-02-25 10:11:08.4180|INFO|ETL|Write data into table PeopleTest processed 63000 records.
2020-02-25 10:11:26.5111|INFO|ETL|Lookup processed 64000 records.
2020-02-25 10:11:26.5359|INFO|ETL|Write data into table PeopleTest processed 64000 records.
2020-02-25 10:11:47.5468|INFO|ETL|Lookup processed 65000 records.
2020-02-25 10:11:47.5720|INFO|ETL|Write data into table PeopleTest processed 65000 records.
2020-02-25 10:12:06.6674|INFO|ETL|Lookup processed 66000 records.
2020-02-25 10:12:06.6901|INFO|ETL|Write data into table PeopleTest processed 66000 records.
2020-02-25 10:12:24.8949|INFO|ETL|Lookup processed 67000 records.
2020-02-25 10:12:24.9238|INFO|ETL|Write data into table PeopleTest processed 67000 records.
2020-02-25 10:12:43.8172|INFO|ETL|Lookup processed 68000 records.
2020-02-25 10:12:43.8417|INFO|ETL|Write data into table PeopleTest processed 68000 records.
2020-02-25 10:13:03.0340|INFO|ETL|Lookup processed 69000 records.
2020-02-25 10:13:03.0984|INFO|ETL|Write data into table PeopleTest processed 69000 records.
2020-02-25 10:13:22.0026|INFO|ETL|Lookup processed 70000 records.
2020-02-25 10:13:22.0301|INFO|ETL|Write data into table PeopleTest processed 70000 records.
2020-02-25 10:13:41.7414|INFO|ETL|Lookup processed 71000 records.
2020-02-25 10:13:41.7688|INFO|ETL|Write data into table PeopleTest processed 71000 records.
2020-02-25 10:14:01.4838|INFO|ETL|Lookup processed 72000 records.
2020-02-25 10:14:01.5122|INFO|ETL|Write data into table PeopleTest processed 72000 records.
2020-02-25 10:14:21.3994|INFO|ETL|Lookup processed 73000 records.
2020-02-25 10:14:21.4345|INFO|ETL|Write data into table PeopleTest processed 73000 records.
2020-02-25 10:14:41.7960|INFO|ETL|Lookup processed 74000 records.
2020-02-25 10:14:41.8231|INFO|ETL|Write data into table PeopleTest processed 74000 records.
2020-02-25 10:15:01.9176|INFO|ETL|Lookup processed 75000 records.
2020-02-25 10:15:01.9452|INFO|ETL|Write data into table PeopleTest processed 75000 records.
2020-02-25 10:15:24.1966|INFO|ETL|Lookup processed 76000 records.
2020-02-25 10:15:24.2231|INFO|ETL|Write data into table PeopleTest processed 76000 records.
2020-02-25 10:15:45.6854|INFO|ETL|Lookup processed 77000 records.
2020-02-25 10:15:45.7139|INFO|ETL|Write data into table PeopleTest processed 77000 records.
2020-02-25 10:16:07.2585|INFO|ETL|Lookup processed 78000 records.
2020-02-25 10:16:07.3351|INFO|ETL|Write data into table PeopleTest processed 78000 records.
2020-02-25 10:16:28.4766|INFO|ETL|Lookup processed 79000 records.
2020-02-25 10:16:28.5123|INFO|ETL|Write data into table PeopleTest processed 79000 records.
2020-02-25 10:16:53.4964|INFO|ETL|Lookup processed 80000 records.
2020-02-25 10:16:53.5268|INFO|ETL|Write data into table PeopleTest processed 80000 records.
2020-02-25 10:17:24.6218|INFO|ETL|Lookup processed 81000 records.
2020-02-25 10:17:24.6590|INFO|ETL|Write data into table PeopleTest processed 81000 records.
2020-02-25 10:17:50.7614|INFO|ETL|Lookup processed 82000 records.
2020-02-25 10:17:50.7907|INFO|ETL|Write data into table PeopleTest processed 82000 records.
2020-02-25 10:18:16.1431|INFO|ETL|Lookup processed 83000 records.
2020-02-25 10:18:16.1769|INFO|ETL|Write data into table PeopleTest processed 83000 records.
2020-02-25 10:18:40.3400|INFO|ETL|Lookup processed 84000 records.
2020-02-25 10:18:40.3701|INFO|ETL|Write data into table PeopleTest processed 84000 records.
2020-02-25 10:19:03.1900|INFO|ETL|Lookup processed 85000 records.
2020-02-25 10:19:03.2187|INFO|ETL|Write data into table PeopleTest processed 85000 records.
2020-02-25 10:19:30.8887|INFO|ETL|Lookup processed 86000 records.
2020-02-25 10:19:30.9223|INFO|ETL|Write data into table PeopleTest processed 86000 records.
2020-02-25 10:19:55.7991|INFO|ETL|Lookup processed 87000 records.
2020-02-25 10:19:55.8336|INFO|ETL|Write data into table PeopleTest processed 87000 records.
2020-02-25 10:20:20.9193|INFO|ETL|Lookup processed 88000 records.
2020-02-25 10:20:20.9493|INFO|ETL|Write data into table PeopleTest processed 88000 records.
2020-02-25 10:20:46.5842|INFO|ETL|Lookup processed 89000 records.
2020-02-25 10:20:46.6260|INFO|ETL|Write data into table PeopleTest processed 89000 records.
2020-02-25 10:21:12.4136|INFO|ETL|Lookup processed 90000 records.
2020-02-25 10:21:12.4904|INFO|ETL|Write data into table PeopleTest processed 90000 records.
```

**Comment by roadrunnerlenny** [comment]
Thanks for the log. That was good information.

I was able to reproduce the error in my test environment - it already takes > 60s to merge ~15.000 records, which is already way too long. The logic itself seems to work as designed, and I can see that it works quick for the first records, but then rapidly slows down the more records are going into the merge.

I haven't found the root cause yet, I suspect it somewhere in the comparison of existing/updated records: https://github.com/roadrunnerlenny/etlbox/blob/5979c3487a91443b02adb5b93ce206e445be99c1/ETLBox/src/Toolbox/DataFlow/DBMerge.cs#L97-L114
I am suspecting it is one of the Linq statements (and perhaps the involved lazy execution), or that I return a reference to a new array. But I wasn't able to put my finger on this - I'll keep you updated.

**Comment by roadrunnerlenny** [comment]
Hi @sliceofbytes ,
I was able to find the performance bottleneck. It was a LINQ query to check if a row does exists and needs to be updated. I replaced it with a Dictionary and a key lookup, which now should work much faster. I will release a new version in the next days containing this fix.

**Comment by roadrunnerlenny** [comment]
Just a quick update: I just released version 1.8.1 which fix this performance issue - it should work a lot faster now: https://www.nuget.org/packages/ETLBox/1.8.1
